reconcile/external_resources/manager.py

```diff
@@ -0,0 +1,330 @@
+                    r.key,
+                )
+                error = True
+        if error:
+            state.resource_status = ResourceStatus.ERROR
+            state.reconciliation_errors = state.reconciliation_errors + 1
+            self.state_mgr.set_external_resource_state(state)
+
+        return need_secret_sync
+
+    def _update_state(self, r: Reconciliation, state: ExternalResourceState) -> None:
+        state.ts = datetime.now(timezone.utc)
+        if r.action == Action.APPLY:
+            state.resource_status = ResourceStatus.IN_PROGRESS
+        elif r.action == Action.DESTROY:
+            state.resource_status = ResourceStatus.DELETE_IN_PROGRESS
+        state.resource_hash = r.resource_hash
+        state.reconciliation = r
+        self.state_mgr.set_external_resource_state(state)
+
+    def _need_secret_sync(
+        self, r: Reconciliation, state: ExternalResourceState
+    ) -> bool:
+        if r.action == Action.APPLY and state.resource_status == ResourceStatus.CREATED:
+            return True
+        return False
+
+    def _sync_secrets(self, keys: Iterable[ExternalResourceKey]) -> None:
+        specs = [spec for key in keys if (spec := self.er_inventory.get(key))]
+        self.secrets_reconciler.sync_secrets(specs=specs)
+
+    def _build_external_resource(
+        self, spec: ExternalResourceSpec, er_inventory: ExternalResourcesInventory
+    ) -> ExternalResource:
+        f = self.factories.get_factory(spec.provision_provider)
+        resource = f.create_external_resource(spec)
+        f.validate_external_resource(resource)
+        return resource
+
+    def _serialize_resource_input(self, resource: ExternalResource) -> str:
+        return json.dumps(
+            resource.dict(exclude={"data": {FLAG_RESOURCE_MANAGED_BY_ERV2}})
+        )
+
+    def handle_resources(self) -> None:
+        desired_r = self._get_desired_objects_reconciliations()
+        deleted_r = self._get_deleted_objects_reconciliations()
+        to_sync: set[ExternalResourceKey] = set()
+        for r in desired_r.union(deleted_r):
+            state = self.state_mgr.get_external_resource_state(r.key)
+            need_sync = self._update_in_progress_state(r, state)
+            if need_sync:
+                to_sync.add(r.key)
+
+            metrics.set_gauge(
+                ExternalResourcesReconcileErrorsGauge(
+                    provision_provider=r.key.provision_provider,
+                    provisioner_name=r.key.provisioner_name,
+                    provider=r.key.provider,
+                    identifier=r.key.identifier,
+                ),
+                float(state.reconciliation_errors),
+            )
+
+            if self._resource_needs_reconciliation(reconciliation=r, state=state):
+                self.reconciler.reconcile_resource(reconciliation=r)
+                self._update_state(r, state)
+
+            if to_sync:
+                self._sync_secrets(keys=to_sync)
```

Comment:
I think you mean to do this after all iterations, right? Wrong indentation?

Reply:
Good catch. Fixed.